Explain that autostart has no start order, and where order and start-delay apply (#505)
xcp-ng/xcp-ng-org: docs/guides/autostart-vm.md

5. Checking the output
`# xe vm-param-list uuid=<VM_UUID> | grep other-config`

## Controlling the boot order and delay

> Two separate VM parameters affect how a VM comes up alongside autostart, and setting one

Reviewer: ```suggestion
Besides autostart, two VM parameters affect how a VM starts. If either is misconfigured, it can make it look as though autostart is broken.
```

Author: Your sentence was better than mine. The paragraph is gone anyway: once I tested the claims under it, the section turned out to be about the wrong feature. `order` and `start-delay` are vApp and HA-failover parameters, and autostart has no ordering at all, so the opening now says that.

incorrectly can look like autostart itself is broken:

- `other-config:start_order=<int>` sets the relative order in which VMs start (lower values
  first).
- `start-delay=<seconds>` controls how long the startup sequence waits **after** this VM

>   before moving on to the next one. Note that two official descriptions exist and they read

Reviewer: To make the text a bit more readable, I suggest adding a line break before _"Note that two official descriptions..."_

Author: Gone with the sentence. I only needed to quote the two descriptions because I had not tested which one was right.

  differently: the CLI reference calls it "the delay to wait before a call to start up the VM
  returns", while the XenAPI field description calls it "the delay to wait before proceeding
  to the next order in the startup sequence". They describe the same blocking behaviour, but

>   the second is the one that matches what you observe, because the delay lands on the *next*

Reviewer 2: "that matches what you observe" ? I think this is your LLM talking to you here, not you talking to users  through our docs.

Reviewer: Besides, I think the sentence itself is quite long and hard to follow.

Author: I tested it. With `start-delay=30`, `xe vm-start` returns after 30 seconds. Cleared, it returns straight away. So the CLI reference wording is the literal one, and the delay applies to the VM it is set on, not the next one. I had it backwards.

I had not tested any of it before. I read the two descriptions and wrote something that made them fit.

The name was wrong too:

```
# xe vm-param-set uuid=<uuid> start-order=1
Error: Unknown field 'start-order'
# xe vm-param-set uuid=<uuid> other-config:start_order=1
# echo $?
0
```

The field is `order`. `other-config:start_order` is accepted and does nothing, other-config takes any key you give it. So my version would have looked like it worked.

Section rewritten from the tests.

Author: This one was more useful than I gave it credit for. The sentence was long because I was trying to make two sources agree without having tested either. Once I tested them, the sentence was not needed at all. Thanks for pushing on it.

>   VM to start rather than on the VM the parameter is set on. A misconfigured `start-delay`

Reviewer: I would also add a line break before _"A misconfigured `start-delay`..."_

Author: Also gone. That was the claim I could not back up, so it is a short note of its own now, with the forum link.

  can also make a VM appear never to autostart, even though the parameter is unrelated to the

>   autostart toggle itself.

Reviewer 2: In what case? Too big a delay set on the VM?

Author: I could not work it out, so the text no longer says it. It comes from one report, [topic 12388](https://xcp-ng.org/forum/topic/12388), where clearing `start-delay` got the VM starting again. I can measure that the delay holds the start call open. I cannot show that this is what stopped the VM starting. The note says that now, with the link.

`/opt/xensource/bin/xapi-autostart-vms` is twenty lines, and this is the part that does the work:

```sh
xe vm-start other-config:auto_poweron=true power-state=halted is-a-template=false is-a-snapshot=false --multiple
```

Everything flagged starts in one call. Nothing reads `order`, so there is no ordering to document, and `order` is not an autostart parameter. The CLI reference had it right: vApp startup and HA failover. The section now says autostart has no order and links to the vApp page, which also answers [topic 12149](https://xcp-ng.org/forum/topic/12149).

One thing I left out of the page, because it is your call. That script logs `auto_poweron is enabled on the pool -- this is an unsupported configuration`, and this guide is what tells people to set the pool flag. I did not find the line in the logs on the host I tested, so all I can say is that the script contains it. I did not want to put that in a docs PR without asking.

5. Checking the output
`# xe vm-param-list uuid=<VM_UUID> | grep other-config`

## 🔢 Start order and delays {#start-order-and-delays}

> Autostart does not define an order. Every flagged VM is started in one go at host boot, so

Reviewer: To make the sentence a bit more natural, I suggest: 

```suggestion
Autostart does not define a startup order. All VMs flagged for autostart are started when the host boots, so
```

Author: @thomas-dkmt Taken as written. Yours reads better than mine.

there is no way to sequence them.

> Ordering and delays belong to [vApps](../vms/vm-lifecycle.md#vapps) instead. Two VM

Reviewer: ```suggestion
To define a startup order or add delays, use [vApps](../vms/vm-lifecycle.md#vapps) instead.
```

Author: @thomas-dkmt Taken, with one adjustment. Your version ends the sentence at "instead.", but the line it replaces ended on "Two VM" and line 48 carries on with "parameters apply when...". Committing the suggestion by itself would leave that line without a subject, so I kept your phrasing and the connective, then rewrapped the paragraph.

parameters apply when an appliance is started, and after an HA failover:

- `order` sets the relative order in which the appliance's VMs start, lower values first.

> - `start-delay` is a number of seconds. A call to start the VM does not return until the

Reviewer: ```suggestion
- `start-delay` specifies a delay in seconds. A call to start the VM does not return until the
```

Author: @thomas-dkmt Applied as written.

  delay has elapsed, so in a sequential appliance start the next VM waits that much longer.

<Terminal title="root@xcp-ng-host — Start order and delays">{`
xe vm-param-set uuid=<VM_UUID> order=1
xe vm-param-set uuid=<VM_UUID> start-delay=30
`}</Terminal>

The parameter is `order`. `start-order` is rejected as an unknown field, and
`other-config:start_order` is accepted without doing anything at all, since `other-config`
holds arbitrary keys.

:::note
`start-delay` has been reported to interfere with autostart: one user found a VM that would
not start at boot until the parameter was cleared
([forum thread](https://xcp-ng.org/forum/topic/12388)). The delay does hold the start call
open, which is measurable, but whether that is what kept the VM from starting has not been

> established. If a VM does not come up after a host reboot, checking `start-delay` costs

Reviewer: "costs nothing" sounds like spoken language than a written documentation, in my opinion. Let me suggest:

```suggestion
If a VM does not come up after a host reboot, checking its start-delay value is a simple troubleshooting step.
```

Author: @thomas-dkmt Applied, with the same line-boundary problem as the vApps one. The sentence ran onto line 69, which is just the word "nothing.", so committing the suggestion alone would have stranded it. I took your sentence and removed the leftover.

One deviation worth flagging: I kept the backticks on `start-delay`, since the parameter is code-formatted everywhere else on the page. Happy to drop them if you would rather it read as plain prose.

5. Checking the output
`# xe vm-param-list uuid=<VM_UUID> | grep other-config`

## 🔢 Start order and delays {#start-order-and-delays}

Autostart does not define a startup order. All VMs flagged for autostart are started when the
host boots, in a single bulk call that never reads the `order` parameter, so you cannot choose
which VM starts first.

> They are started one after another rather than all at once, so a `start-delay` on one VM does

Reviewer: I think we can afford to start this new paragraph by stating the subject ('VMs') again:

```suggestion
VMs are started one after another rather than all at once, so a `start-delay` on one VM does
```

  delay has elapsed, so in a sequential appliance start the next VM waits that much longer. The
  VM it is set on is already running by then. It is the call that waits, not the VM.

<Terminal title="root@xcp-ng-host — Start order and delays">{`
xe vm-param-set uuid=<VM_UUID> order=1
xe vm-param-set uuid=<VM_UUID> start-delay=30
`}</Terminal>

The parameter is `order`. `start-order` is rejected as an unknown field, and
`other-config:start_order` is accepted without doing anything at all, since `other-config`
holds arbitrary keys.

:::warning
An appliance is not started when the host boots. Nothing in the boot sequence calls
`appliance-start`, so grouping VMs into an appliance does not on its own bring them up after a
reboot. They also need `other-config:auto_poweron`, and once they have it they take the
autostart path above, which ignores `order`. Start an appliance with `xe appliance-start`, or
let HA restart it after a failover.
:::

:::note
`start-delay` has been reported to interfere with autostart: one user found a VM that would
not start at boot until the parameter was cleared
([forum thread](https://xcp-ng.org/forum/topic/12388)). A delayed start does hold up the VMs

> queued behind it, which explains one coming up late, but not one that never comes up at all.

Reviewer: _"but not one that never comes up at all"_ introduces a double negative, which can be confusing. Can we avoid it while keeping the sentence clear?

> That part is unexplained. If a VM does not appear after a host reboot, checking its

Reviewer: To make the sentence more fluid, I suggest a little reordering:

_"If a VM does not appear after a host reboot, it is worth checking its `start-delay` value and waiting to see whether it is simply queued behind another VM's delay."_

5. Checking the output
`# xe vm-param-list uuid=<VM_UUID> | grep other-config`

> ## 🔢 Start order and delays {#start-order-and-delays}

Reviewer: Since we're no longer using emojis in section headers (as per https://github.com/xcp-ng/xcp-ng-org/pull/525 which has just been approved and merged), this emoji should be removed as well:

```suggestion
## Start order and delays {#start-order-and-delays}
```

Author: Applied in `67eba01`. Heading text and anchor untouched, so no links break.

Only the heading this PR adds needed it. The two above it are pre-existing lines the diff never touches, so #525 already stripped them, and there is nothing here to put them back.
